pkg/kvstore/etcd_test.go

```diff
@@ -1664,3 +1665,183 @@ func TestShuffleEndpoints(t *testing.T) {
 		t.Errorf("Shuffle() did not modify s2 in 10 retries")
 	}
 }
+
+type EtcdRateLimiterSuite struct {
+	etcdClient *etcdAPI.Client
+	maxQPS     int
+}
+
+var _ = Suite(&EtcdRateLimiterSuite{})
+
+func (e *EtcdRateLimiterSuite) setupWithRateLimiter() {
+	// The rate limiter is configured with max QPS and burst both
+	// configured to the provided value for rate limit option.
+	setupDummyWithConfigOpts("etcd", map[string]string{
+		EtcdRateLimitOption: fmt.Sprintf("%d", e.maxQPS),
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+	<-Client().Connected(ctx)
+}
+
+func (e *EtcdRateLimiterSuite) setupWithoutRateLimiter() {
+	SetupDummy("etcd")
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+	<-Client().Connected(ctx)
+}
+
+func (e *EtcdRateLimiterSuite) SetUpSuite(c *C) {
+	e.maxQPS = 3
+
+	// setup client
+	cfg := etcdAPI.Config{}
+	cfg.Endpoints = []string{etcdDummyAddress}
+	cfg.DialTimeout = 0
+	cli, err := etcdAPI.New(cfg)
+	c.Assert(err, IsNil)
+	e.etcdClient = cli
+}
+
+func (e *EtcdRateLimiterSuite) TearDownSuite(c *C) {
+	err := e.etcdClient.Close()
+	c.Assert(err, IsNil)
+}
+
+func (e *EtcdRateLimiterSuite) getKey(prefix string, c int) string {
+	if prefix != "" {
+		return fmt.Sprintf("%s-%d", prefix, c)
+	}
+
+	return fmt.Sprintf("foobar-%d", c)
+}
+
+func (e *EtcdRateLimiterSuite) putRandomKeys(c *C, prefix string, count int) {
+	for i := 0; i < count; i++ {
+		key := e.getKey(prefix, i)
+
+		_, err := e.etcdClient.Put(context.Background(), key, "bar")
+		c.Assert(err, IsNil)
+	}
+}
+
+func (e *EtcdRateLimiterSuite) deleteRandomKeys(c *C, prefix string, count int) {
+	for i := 0; i < count; i++ {
+		key := e.getKey(prefix, i)
+
+		_, err := e.etcdClient.Delete(context.Background(), key)
+		c.Assert(err, IsNil)
+	}
+}
+
+func (e *EtcdRateLimiterSuite) TestRateLimiterGetOps(c *C) {
+	randomPath := c.MkDir()
+	txnCount := 15
```

Comment:
We don't need to make it 15, it can be `e.maxQPS+2`. That should return the same results as well.

Reply:
We will at least need this to be `e.maxQPS * 2 + 2`.
If we use e.maxQPS + 2 the operation can be performed in less than a second because the rate limiter comes with a granularity of a second and the tokens in the rate limiter are regenerated every second.